solutions/gold/usaco-1234.mdx

```diff
@@ -0,0 +1,185 @@
+
+<Spoiler title="Hint">
+
+Think about the [Commutative Property](https://en.wikipedia.org/wiki/Commutative_property)!
+
+</Spoiler>
+
+<Spoiler title="Detailed Explanation">
+
+**Note**: This explanation presents the solution in a very abstract way, but it should be mostly intuitive!
+
+Consider the *properties* of an expression after an instruction is applied to it:
+- $*d$: all coefficients must be a multiple of $d$
+- $+x$: the coefficient of x is *exactly* $1$
+Let's call these the **traits** of that instruction, and let's say an instruction X is **violated** by a subsequent instruction Y if, after applying instruction Y, the traits of instruction X no longer hold. As an example, let instruction X be $*2$ and instruction Y be $+y$. After we apply instruction X to the expression $x$, we get $2x$, which satisfies the trait that all coefficients must be a multiple of 2. However, after applying instruction Y, we get $2x + y$, an expression whose coefficients are no longer all even; therefore, instruction Y violates instruction X. From these defintions, we get an important fact: **two operations that don't violate each other must be commutative** (try to see the intuition)!
+
+If an expression is contained in both sets $A$ and $B$, by definition it can be written as both $E(i - 1, j) \rightarrow a_i$ and $E(i, j - 1) \rightarrow b_j$, and therefore **must satisfy the traits of both $a_i$ and $b_j$**. Thus, we have that:
+1. $a_i$ and $b_j$ are commutative
+2. the expression can be written as $\text{[some expression]} \rightarrow a_i \rightarrow b_j$ (or $\text{[some expression]} \rightarrow b_j \rightarrow a_i$).
+However, because this expression must still be in $I(i, j)$, we can see that $\text{[some expression]}$ must be in $I(i - 1, j - 1)$, as we've already placed both $a_i$ and $b_j$ at the end of the instruction sequence. Therefore, we have that $|A \cap B| = |I(i - 1, j - 1)|$ when $a_i$ and $b_j$ are commutative, and $0$ otherwise.
+
+</Spoiler>
+
+<Spoiler title="TL;DR">
+
+If $a_i$ and $b_j$ are commutative operations (i.e. $\rightarrow a_i \rightarrow b_j$ does the same thing as $\rightarrow b_j \rightarrow a_i$), we overcount the set $I(i - 1, j - 1) \rightarrow a_i \rightarrow b_j$. Make sure to remember the edge case of the $0$ expression as well (detailed more in code later)!
+
+</Spoiler>
+
+## Implementation
+
+**Time Complexity**: $\mathcal{O}(N^2)$
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <iostream>
+using namespace std;
+
+using ll = long long;
+
+const int N = 2e3 + 1;
+const int MOD = 1e9 + 7;
+
+vector<char> a, b;
+
+// zero[i][j] -> whether 0 is in I(i, j)
+vector<vector<bool>> zero;
+
+// dp[i][j] -> number of nonzero expressions in I(i, j)
+vector<vector<ll>> dp;
+bool digit(char c) { return c >= '1' && c <= '9'; }
```

Comment:
actually i think there's a built-in func for this
https://en.cppreference.com/w/cpp/string/byte/isdigit